### PR TITLE
[QAOA (Secondary)] Adds Bit-Flip Mixer

### PR DESCRIPTION
**Pull request by Lucaman99**
**Context:**

Part of the ongoing effort to add QAOA functionality to PennyLane

**Description of the Change:**

Adds a new bit-flip mixer, which is useful when performing constrained QAOA for MaxIndependentSet, MinVertexCover, MaxClique, etc.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by Lucaman99** [comment]
@josh146 This PR is ready for review! Since it's fairly short, I think we probably only need two reviewers?

**Comment by josh146** [comment]
Great, thanks @Lucaman99! I've tagged @ixfoduap and @trbromley for review 👍

**Comment by Lucaman99** [comment]
Thanks @josh146!

File: .github/CHANGELOG.md
Comment (trbromley) [inline .github/CHANGELOG.md:7]: More out of curiosity - what makes this a new feature vs an improvement?

File: pennylane/qaoa/mixers.py
Comment (trbromley) [inline pennylane/qaoa/mixers.py:125]: Minor thing: how about `N(v)` instead of `nbhd(v)`? This is what is used here: https://en.wikipedia.org/wiki/Neighbourhood_(graph_theory)

File: pennylane/qaoa/mixers.py
Comment (trbromley) [inline pennylane/qaoa/mixers.py:124]:
```suggestion
    .. math:: H_M \ = \ \displaystyle\sum_{v \in V(G)} \frac{1}{2^{d(v)}} X_{v}
```
image

File: pennylane/qaoa/mixers.py
Comment (trbromley) [inline pennylane/qaoa/mixers.py:138]: Minor thing - good to be consistent with the docstring of `xy_mixer()`

File: pennylane/qaoa/mixers.py
Comment (trbromley) [inline pennylane/qaoa/mixers.py:140]: I found this one less clear to understand, to me it looks like the choice of n determines whether we plus or minus Z_w. But this is not a bit flip?

Comment (Lucaman99) [reply pennylane/qaoa/mixers.py:140]: In the case that n = 0, we have a product over (1 + Z) acting on neighbouring nodes. This will be 0 if there is some neighbouring wire that is labelled with 1 (since (1 + Z|1>) = (1 - 1) = 0), so the state of vertex v will not be flipped by the X_v operation unless all neighbouring nodes are in the |0> state. In the other case, we have (1 - Z) acting on all neighbouring nodes, so all neighbouring nodes have to be in the |1> state in order for v to be flipped.

File: .github/CHANGELOG.md
Comment (Lucaman99) [reply .github/CHANGELOG.md:18]: Yeah, I think it could be either. However, since this is technically an addition to the pre-existing collection of mixers, I can see how it would be an improvement 👍

File: pennylane/qaoa/mixers.py
Comment (trbromley) [inline pennylane/qaoa/mixers.py:140]:
```suggestion
         n (int): Either :math:`0` or :math:`1`. When :math:`n=0`, a bit flip is performed on
             vertex :math:`v` only when all neighbouring nodes are in state :math:`|0\rangle`. 
             Instead, for :math:`n=1`, a bit flip is performed only when all the neighbours of 
             :math:`v` are in the state :math:`|1\rangle`.
```
Thanks for the clarification! This suggestion is longer, but I think captures what you said clearly.

Comment (trbromley) [reply pennylane/qaoa/mixers.py:140]: image

**Review by trbromley — APPROVED** [review]
Thanks @Lucaman99! Just left a final suggestion for a docstring clarification

**Comment by Lucaman99** [comment]
Thanks @trbromley!

**Comment by josh146** [comment]
@ixfoduap I'm going to remove myself from review for this one - 2 reviewers should be sufficient for a PR of this magnitude 🙂 

Feel free to comment and tag me on any specifics you would like me to look at though

**Comment by josh146** [comment]
Hey @Lucaman99, are we just waiting for @ixfoduap's review for this PR before it can be merged in?

**Comment by Lucaman99** [comment]
@josh146 Yep, that's right!

File: pennylane/qaoa/mixers.py
Comment (ixfoduap) [inline pennylane/qaoa/mixers.py:130]:
```suggestion
    of vertex :math:`v`. In addition, :math:`\mathbb{I}` is the identity operator, and :math:`X_v`, :math:`Z_v`
```
to avoid too many "ands"

**Review by ixfoduap — APPROVED** [review]
Looks great!

**Comment by Lucaman99** [comment]
Awesome! 🎉

**Pull request merged**
